### PR TITLE
✅ amp-recaptcha-input: Added support for integration test mocking in express server

### PR DESCRIPTION
**Pull request by torch2424**
depends on #20539
relates to #10059
relates to #2273

This does the following:

* Sets up a few paths on the express server to intercept calls to the recaptcha frame, to substitute the recaptcha API url.
* Allows the recaptcha service to get the testLocation provided by integration tests, to set the iframe src correctly.

screen shot 2019-01-25 at 5 26 33 pm

### CODE REVIEW

File: extensions/amp-recaptcha-input/0.1/amp-recaptcha-service.js [hunk 1]
```diff
@@ -251,10 +251,17 @@ export class AmpRecaptchaService {
    */
   getRecaptchaFrameSrc_() {
     if (getMode().localDev || getMode().test) {
-      return ampToolboxCacheUrl.createCurlsSubdomain(this.win_.location.href)
+      let winLocation = this.win_.location;
+      if (this.win_.testLocation) {
```
Comment (zhouyx) [inline extensions/amp-recaptcha-input/0.1/amp-recaptcha-service.js]: When will `this.win_.testLocation` be set?

File: build-system/recaptcha.test.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+window.grecaptcha = {
+  ready: (callback) => callback(),
+  execute: () => Promise.resolve('recaptcha-mock')
+};
+`;
+
+const recaptchaFrameRequestHandler = (req, res, next) => {
+  if (global.AMP_TESTING) {
+    fs.readFileAsync(pc.cwd() + req.path, 'utf8').then(file => {
+      file = file.replace(/initRecaptcha\(.*\)/g, 'initRecaptcha("/recaptcha/mock.js?sitekey=")');
```
Comment (zhouyx) [inline build-system/recaptcha.test.js]: 👍 great idea!

Comment (torch2424) [reply build-system/recaptcha.test.js]: Thanks! 😄

File: extensions/amp-recaptcha-input/0.1/amp-recaptcha-service.js (hunk 1 above)
Comment (torch2424) [reply extensions/amp-recaptcha-input/0.1/amp-recaptcha-service.js]: It is set in: https://github.com/ampproject/amphtml/blob/master/testing/iframe.js#L100

😄  Let me know what you think!

File: build-system/app.js [hunk 3]
```diff
@@ -118,6 +125,25 @@ if (!global.AMP_TESTING) {
   });
 }
 
+/*
+ * Intercept Recaptcha frame for,
+ * integration tests. Using this to mock
+ * out the recaptcha api.
+ */
+app.get(
+    '/dist.3p/current/recaptcha.max.html',
```
[the remark this answers is not in the document]
Comment (torch2424) [reply build-system/app.js]: Good point! Thanks!

File: build-system/app.js [hunk 4]
```diff
@@ -118,6 +125,25 @@ if (!global.AMP_TESTING) {
   });
 }
 
+/*
+ * Intercept Recaptcha frame for,
+ * integration tests. Using this to mock
+ * out the recaptcha api.
+ */
+app.get(
+    '/dist.3p/current/recaptcha.*html',
```
[the remark this answers is not in the document]
Comment (torch2424) [reply build-system/app.js]: Yes, you are correct there are multiple `current*` directories. So I just used a blob here as well 😄

File: extensions/amp-recaptcha-input/0.1/amp-recaptcha-service.js (hunk 1 above)
Comment (lannka) [reply extensions/amp-recaptcha-input/0.1/amp-recaptcha-service.js]: I'm wondering what's the real value of `this.win_.location` and `this.win_.testLocation` in test mode. Should them both be `http://localhost:9876`?

**Comment by torch2424** [comment]
@lannka Made requested changes 😄

**Comment by lannka** [comment]
@torch2424 I guess you missed my one last comment regarding to the testLocation.

I want to see if we can remove it.

**Pull request merged**